### ISSUE TITLE
Two schemas are referenced with the same name but different content. Renamed Item to Item-2.

### ISSUE THREAD

**Issue by colinmollenhour** [Redocly/openapi-cli#661]
If using subdirectories to organize a large number of components and some of these components share the same file basename (but exist in different subdirectories) you get warnings like this:

```
[6] paths/beta/rating/merchantRates.yaml:59:7 at #/get/parameters/3

Two schemas are referenced with the same name but different content. Renamed fields to fields-5.
```

This also comes up with models like `models/Order/Item.yaml` and `models/Shipment/Item.yaml` because both files are named `Item` so they are merged in as `Item` and `Item-2` and so-on.

**Example spec**

`openapi.yaml`
```
openapi: 3.0.3
info:
  title: test
paths:
  '/api/global/beta/rating/rateGroups':
    $ref: 'paths/beta/rating/rateGroups.yaml'
  '/api/global/beta/rating/merchantRates':
    $ref: 'paths/beta/rating/merchantRates.yaml'
```

`paths/beta/rating/rateGroups.yaml`
```
get:
  summary: List all Rate Groups
  description: foo
  parameters:
    - $ref: ../../../components/parameters/rate/group/fields.yaml
```

`paths/beta/rating/merchantRates.yaml`
```
get:
  summary: List all Merchant Rates
  description: foo
  parameters:
    - $ref: ../../../components/parameters/merchant/rate/fields.yaml
```

**Expected behavior**
The refs are valid so ideally there should not be such warnings and ideally it should not require the refs to be declared explicitly in the top level file as a workaround as that is much more tedious than just directly referencing the files where they are needed.

**`openapi-cli` Version(s)**
1.0.0-beta.73

**Comment by tatomyr** [comment]
Hi @colinmollenhour,
Thank you for reporting. I agree it might feel like the behaviour is not too obvious. We might consider implementing this in future. 
However, you can safely skip these warnings, I believe. Or alternatively you can also try using different file names.

**Comment by adamaltman** [comment]
These warnings are for your benefit (not because there is an issue with your definitions) -- they could be posed as "informational" instead of "warning". It is letting you know we cannot use the same file name when we move the reference from a remote reference to the components section in the bundled file. If you want it to be completely dereferenced, you could use the dereferenced flag.

However, I think there could be a better way for us to come up with schema names. For example, in the case of a conflict, we append `-2`, `-3`, `-etc.` In your example above, we could try something like using the folder name such as `Shipment_Item`.

**Comment by colinmollenhour** [comment]
I like them remaining as references as that keeps the spec concise and agree that the warnings can safely be ignored but then other warnings get masked quickly.. Using one parent folder name as a prefix would be a big improvement in my opinion and probably work in 95% of cases. As of now I've resorted to naming files like `Order/OrderItem.yaml` instead of `Order/Item.yaml` to avoid the warnings.

**Comment by adamaltman** [comment]
The reverse behavior could be `strict` mode.

**Comment by djMax** [comment]
I'm getting this error when the refs point to the SAME file but from different base paths - i.e. if I have `$ref: something/file.yaml` and in the `something` directory I have `$ref: file.yaml` it will complain.

**Comment by troccoli** [comment]
Is there a specific rule that can be turned off for this?

**Comment by tatomyr** [comment]
@troccoli do you mean turning off the warnings? No, currently there's no way to suppress them.

**Comment by troccoli** [comment]
Yes @tatomyr that's what I meant.

I'm fine not caring for those warnings but it would be nice if they can be excluded from the output so that I don't miss other warnings.

Thanks for answering anyway

**Comment by kennethaasan** [comment]
Why is not the `title` in the schema used instead of the file name?

**Comment by domenico-renna** [comment]
This behavior is really annoying, we have a spec-root.yaml file which is referenced in many other spec files for common answers or parameters also we use a folder related to every resource containg a spec.yaml (with the routes specification) and a schema.yaml file (with the entity obyect definition).

With this behaviour you full our console with caothic and unusefull log hiding real error if occurs

**Comment by lornajane** [comment]
@domenico-renna Could you open a new issue with some examples for that use case? It sounds a little different to the example above with individual schemas in same-named files. If you can link to this issue, that would help.

**Comment by lornajane** [comment]
One approach to solving this is to name the files more meaningfully so that those names get used when bundling. The simplest way to do this (that I use whenever I see this problem) is to bundle the API description, rename everything the way you want, then use the `split` command (see https://redocly.com/docs/cli/commands/split/) to turn it back into a structure of manageable-sized files.

If we were to implement changes, it could be to improve the naming policy so that the tool automatically picks probably-sensible names and doesn't warn about adding suffixes. Adam's suggestion further up the thread of adding directory prefixes would work for most of the use cases here I think. The first item encountered would still get its current name (e.g. `fields` in the original issue example), but the next one would get a path-related prefix (e.g. `rate-fields`) to distinguish them more meaningfully.

**Comment by dchertousov** [comment]
I have similar issue.

I have created directory structure for the params, so it will be easily understandable where they come from. And there is one param, that could be passed either in path or in query, depending on the call, so the have the same file name though in different dirs (`path/application-id.yaml` and `query/application-id.yaml`):

```
└── openapi
    ├── parameters
    │   ├── _index.yaml
    │   ├── header
    │   │   └── x-api-key.yaml
    │   ├── path
    │   │   ├── application-id.yaml
    │   │   ├── organization-id.yaml
    │   │   ├── payment-id.yaml
    │   │   └── subscription-id.yaml
    │   └── query
    │       ├── application-id.yaml
    │       ├── date.yaml
    │       └── return-expired-subscriptions.yaml
```

May be it is possible to take not only the file name, but the whole relative path to account, so they will be different?

**Comment by mlautenschlager-tc** [comment]
I would also vote for an option to turn it off. It's spamming my log. 

why don't you add the path to the component name like?

**/ssl/request.yaml = ssl_request**

**Comment by onavratil-monetplus** [comment]
Having two models with same name in a different directory context is no sin. This breaks my pipeline and there is no way to suppress the warning. If nothing else, why isn't it on stderr...
Here we go creating temporary files just to strip a warning.

**Comment by tatomyr** [comment]
@onavratil-monetplus it should go to the stderr. Thanks for noticing this bug. 

> This breaks my pipeline and there is no way to suppress the warning.

So the warning itself is what breaks your pipeline? 

If yes, I'll try to fix that today.

**Comment by onavratil-monetplus** [comment]
Yes, because i have a chain of other tools linked to the output of `redocly bundle`.
Now, as you can imagine, if the dereferenced YAML content is suffixed with a textual warning, it breaks the format structure and subsequent tooling yields cryptic error logs (I understand that this is out of scope of redocly, but anyway)
```
WARNING: Cannot parse or dereference YAML /docs/test-api.yaml: while scanning a simple key
  in "<byte string>", line 258, column 1:
    [1] test-api.yaml:75:21 at #/pat ... 
    ^
could not find expected ':'
  in "<byte string>", line 260, column 1:
    Two schemas are referenced with  ...
```
In any case, the warning content being questionable is one thing, but I cannot wrap my head around the fact that the warning log goes to `stdout` and not `stderr`.
I mean, even if it was useful and informative, stdout makes less sense. Especially considering that the "usual" redocly log goes to stderr
```
$ redocly bundle test-api.yaml 1>/dev/null
bundling test-api.yaml...
(node:55752) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
📦 Created a bundle for test-api.yaml at stdout 36ms.
```
And that holds even if -o parameter is used.
So basically, successful log goes to `stderr`, warnings go to `stdout`...

**Comment by onavratil-monetplus** [comment]
I made a workaround where I let redocly -o the bundled yaml into some tempfile and have the subsequent tooling read the contents of the file instead of stdout of redocly. Yet its rather...impractical.

**Comment by tatomyr** [comment]
Added a new `--component-renaming-conflicts-severity` option for the `bundle` command to control how naming conflicts are reported when bundling API descriptions with external references. Please see the [docs](https://redocly.com/docs/cli/commands/bundle#configure-component-renaming-conflicts).
This should cover the original @colinmollenhour 's issue, and the opposite behaviour (@pvonr, FYI).

The option is available with [v2.16.0](https://redocly.com/docs/cli/changelog#2160-2026-02-06).
Also, this release includes changes to where the bundling problems go: instead of stdout, they are now being redirected to stderr (@onavratil-monetplus, could you verify it works as expected now?).

**Comment by onavratil-monetplus** [comment]
Hi, thanks for the patch. Sorry for late reply, am out of country.
Tried with `2.20.0`

- [X]  `$ redocly bundle test-api.yaml` produces warning on `stderr` and dereferenced OAS output on `stdout`. Redirecting `2>/dev/null` effectively removes the warning from terminal window.
- [X]  `$ redocly bundle test-api.yaml --component-renaming-conflicts-severity off` effectively suppresses generation of the warning as such.

Works like a charm. Going to remove my tempfile workaround.
